File: codechef/contests/JAN14/FGFS.py
from collections import defaultdict
import sys

class activity:
	def __init__(self, start, end, comp):
		self.start = start
		self.end = end
		self.comp = comp

	def __repr__(self):
		string = "%d %d %d"%(self.start, self.end, self.comp)
		return string

def compare(act1, act2):
	if act1.end == act2.end:
		return -1 if act1.start < act2.start else 0
	return -1 if act1.end < act2.end else 0

def solve(n, k, a):
	ans = 0
	a.sort(compare)
	d = defaultdict(int)
	for act in a:
		start = act.start
		end = act.end
		comp = act.comp
		if d[comp] == 0:
			d[comp] = end
			ans += 1
		else:
			if start >= d[comp]:
				ans += 1
				d[comp] = end
	return ans
inputstring = sys.stdin.read()
inputarray = inputstring.split("\n")
t = int(inputarray[0])
ii = 1
ans = ""
for _ in xrange(t):
	n, k = map(int, inputarray[ii].split())
	ii += 1
	a = []
	for i in xrange(n):
		a.append(activity(*map(int, inputarray[ii].split())))
		ii += 1
	ans += `solve(n, k, a)` +"\n"

sys.stdout.write(ans)

File: codechef/contests/JAN14/FGFS.cpp
#include <cstdio>
#include <algorithm>
#include <map>

using namespace std;

class activity{
public:
	int start, end, comp;
	activity(){}
	activity(int s, int e, int c){
		start = s;
		end = e;
		comp = c;
	}
};

bool compare(activity act1, activity act2){
	if (act1.end == act2.end){
		return act1.start < act2.start ? true : false;
	}
	return act1.end < act2.end ? true : false;
}

int solve(int n, int k, activity a[]){
	int ans = 0;
	sort(a, a+n, compare);
	map<int, int> d;
	for(int i = 0; i < n;i += 1){
		activity act = a[i];
		if (d[act.comp] == 0){
			d[act.comp] = act.end;
			ans += 1;
		}else{
			if (act.start >= d[act.comp]){
				ans += 1;
				d[act.comp] = act.end;
			}
		}
	}
	return ans;
}

int main(void){
	int t, n, k, x, y, z;
	scanf("%d",&t);
	while (t--){
		scanf("%d %d", &n, &k);
		activity a[n];
		for (int i = 0; i < n; i+=1){
			scanf("%d %d %d", &x, &y, &z);
			a[i] = activity(x, y, z);
		}
		printf("%d\n",solve(n, k, a));
	}
}